File: jianzhioffer/firstround/jianzhi_12.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        def dfs(i, j, k):
            if not 0 <= i <= len(board)-1 or not 0 <= j <= len(board[0])-1 or board[i][j] != word[k]:
                return False
            if k == len(word) - 1:
                return True
            board[i][j] = ''
            res = dfs(i+1, j, k+1) or dfs(i-1, j, k+1) or dfs(i, j-1, k+1) or dfs(i, j+1, k+1)
            board[i][j] = word[k]
            return res

        for row in range(len(board)):
            for col in range(len(board[0])):
                if dfs(row, col, 0):
                    return True
        return False


if __name__ == '__main__':
    board = [["C","A","A"],["A","A","A"],["B","C","D"]]
    word = "AAB"
    obj = Solution()
    result = obj.exist(board, word)
    print(result)

File: jianzhioffer/secondround/jianzhi_12.cpp
#include <vector>
#include <string>
using namespace std;

//DFS+剪枝
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        if(word.empty() || board.empty())
            return false;
        rows = board.size();
        cols = board[0].size();
        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                if (dfs(board, word, i, j, 0))
                    return true;
            }
        }
        return false;
    }

private:
    int rows, cols;
    bool dfs(vector<vector<char>>& board, string& word, int i, int j, int k){
        if (i>=rows || i<0 || j>=cols || j<0 || board[i][j]!=word[k])
            return false;
        else if (k==word.size()-1)
            return true;
        board[i][j] = '\0';
        bool res = dfs(board, word, i-1, j, k+1) || dfs(board, word, i, j-1, k+1) || 
                   dfs(board, word, i, j+1, k+1) || dfs(board, word, i+1, j, k+1);
        board[i][j] = word[k];
        return res;
    }
};

int main(int argc, char const *argv[])
{
    int m = 3;
    int n = 4;
    vector<vector<char>> nums(m, vector<char>(n));

    nums[0] = {'A','B','C','E'};
    nums[1] = {'S','F','C','S'},
    nums[2] = {'A','D','E','E'};
    string w = "ABCCED";
    Solution obj;
    bool result = obj.exist(nums, w);
    return 0;
}
